pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go

```diff
@@ -54,13 +55,84 @@ func (ac *OSSAccessControlService) getUsageMetrics() interface{} {
 	return 1
 }
 
+func (ac *OSSAccessControlService) saveFixedRole(role accesscontrol.RoleDTO) error {
+	accesscontrol.FixedRolesMutex.Lock()
+	defer accesscontrol.FixedRolesMutex.Unlock()
+
+	if storedRole, ok := accesscontrol.FixedRoles[role.Name]; ok {
+		if storedRole.Version >= role.Version {
+			return accesscontrol.ErrVersionLE
+		}
+	}
+	// Save role
+	accesscontrol.FixedRoles[role.Name] = role
+
+	return nil
+}
+
+func (ac *OSSAccessControlService) assignFixedRole(role accesscontrol.RoleDTO, builtInRoles []string) error {
+	accesscontrol.FixedRoleGrantsMutex.Lock()
+	defer accesscontrol.FixedRoleGrantsMutex.Unlock()
+
+	for _, builtInRole := range builtInRoles {
+		assignments := []string{}
+		if assignments, ok := accesscontrol.FixedRoleGrants[builtInRole]; ok {
+			alreadyAssigned := false
+			for _, assignedRole := range assignments {
+				if assignedRole == role.Name {
+					alreadyAssigned = true
+				}
+			}
+			if alreadyAssigned {
+				return accesscontrol.ErrBuiltinRoleAlreadyAdded
+			}
+		}
+		assignments = append(assignments, role.Name)
+		accesscontrol.FixedRoleGrants[builtInRole] = assignments
+	}
+	return nil
+}
+
+func (ac *OSSAccessControlService) validateFixedRole(role accesscontrol.RoleDTO) error {
+	if !strings.HasPrefix(role.Name, accesscontrol.FixedRolePrefix) {
+		return accesscontrol.ErrFixedRolePrefixMissing
+	}
+	return nil
+}
+
+// RegisterFixedRole saves a fixed role and assigns it to built-in roles
+func (ac *OSSAccessControlService) RegisterFixedRole(_ context.Context, role accesscontrol.RoleDTO, builtInRoles ...string) error {
+	err := ac.validateFixedRole(role)
+	if err != nil {
+		return err
+	}
+
+	err = ac.saveFixedRole(role)
+	if err != nil {
+		return err
+	}
+
+	err = ac.assignFixedRole(role, builtInRoles)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
 // Evaluate evaluates access to the given resource
 func (ac *OSSAccessControlService) Evaluate(ctx context.Context, user *models.SignedInUser, permission string, scope ...string) (bool, error) {
 	return evaluator.Evaluate(ctx, ac, user, permission, scope...)
 }
 
 // GetUserPermissions returns user permissions based on built-in roles
 func (ac *OSSAccessControlService) GetUserPermissions(ctx context.Context, user *models.SignedInUser) ([]*accesscontrol.Permission, error) {
+	// Lock fixed role grants and list for reading
+	accesscontrol.FixedRoleGrantsMutex.RLock()
+	defer accesscontrol.FixedRoleGrantsMutex.RUnlock()
```

Comment:
Shouldn't we lock/unlock exactly when we access a map? Looks like map will be locked during the all function execution time.